acasclient/selfile.py

```diff
@@ -0,0 +1,851 @@
+        }
+    )
+    # Datatype need to be provided as a dict (or ordereddict)
+    expt.calculated_results_datatype = {
+        'Corporate ID': 'Datatype',
+        'IC50 (nM)': 'Numeric',
+        'Hill Slope': 'Numeric',
+        'Assay Comment': 'Text'
+    }
+    expt.saveAs('my_expt.csv')
+    """
+
+    def __init__(self, file_name=None):
+        """
+        :param file_name: Path to the ACAS Generic Experiment file to load.
+        :type file_name:  str
+        """
+
+        super().__init__(file_name)
+        self._format = GENERIC
+
+    def _parse(self, read_func):
+        raw_df = read_func(
+            self.file_name,
+            header=None
+        )
+        self._parseMeta(raw_df.iloc[0:self._meta_rows, [0,1]])
+
+        # Slice the 'calculated results'.
+        self._expt_df = raw_df.iloc[self.RESULTS_HEADER['Datatype']+1:, :]
+        self._expt_df.columns = self._expt_df.iloc[0]
+        self._expt_df = self._expt_df.drop(self._expt_df.index[0])
+
+        # Do datatype after determining expt_df so the df.columns can be zipped.
+        dt_rows = [self.RESULTS_HEADER['Datatype'], self.RESULTS_HEADER['Datatype']+1]
+        datatype = raw_df.iloc[dt_rows, :]
+        self._datatype = self._parseDataType(datatype)
+
+        # Change tracking vars.
+        self._expt_has_changed = False
+        self._meta_has_changed = False
+        self._datatype_has_changed = False
+
+    def _parseMeta(self, meta_df):
+        meta = collections.OrderedDict([(x, '') for x in self.META_HEADER])
+        for idx, row in meta_df.iterrows():
+            if row[0] in meta:
+                meta[row[0]] = row[1]
+            if row[0] in self.RESULTS_HEADER:
+                self.RESULTS_HEADER[row[0]] = idx
+        self._experimental_meta_data = meta['Experiment Meta Data']
+        self._format = meta['Format']
+        self._protocol_name = meta['Protocol Name']
+        self._experiment_name = meta['Experiment Name']
+        self._scientist = meta['Scientist']
+        self._notebook = meta['Notebook']
+        self._page = meta['Page']
+        self._assay_date = meta['Assay Date']
+        self._project = meta['Project']
+
+    def _parseDataType(self, datatype):
+        datatype = list(datatype.T.to_dict().values())
+        datatype = list(datatype[0].values())
+        datatype = collections.OrderedDict(zip(self.expt_df.columns, datatype))
+        return datatype
+
+    def saveAs(self, file_name=None, file_type=None):
+        """
+        :param file_name: Path to output file to save record.
+        :type file_name: str
+        """
+
+        # Must provide either a file_name or a file_type but not both.
+        if file_name and file_type:
+            raise ValueError('Must provide either a file_name or a file_type but not both.')
+
+        # If file_name is provided, use it.
+        if file_name:
+            file_type = get_file_type(file_name)
```

Comment:
Does this need and 'else' clause?

Reply:
sorry forgot to note this change to saveAs.  I needed a method to get the file as text returned so I added the file_type as a parameter.  This shouldn't need an else because above we test if file_name and file_type are provided and throw an error.  Later file_name which is None causes the `df_out.to_csv` and `df_out.to_excel` to return their output instead of writing to a file.

Reply:
Moving this PR to a draft as there are some more changes I want to make to the saveAs.  I realize my changes would break in the case of asking for an excel file_type as to_excel requires a file like object.

Reply:
Would df_out.to_csv() work, writing to a stringIo object as your file handle?